### PR TITLE
2034: Add LogHandler for logging REST API calls at server side

### LINKED ISSUE #2034: `--log-level=debug` should log REST API calls

### PR DESCRIPTION
**Pull request by brb**
Fixes #2034

### CODE REVIEW

**Comment by brb** [comment]
NB: similar handler exists in the scope project. Probably, at some point it will make sense to create a separate repo for code which can be reused by both projects.

**Comment by bboreham** [comment]
Please take the unnecessary change out of prog/weaver/http.go

File: common/http.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// ServeHTTP implements the http.Handler interface.
+func (l *logHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	writerWrapper := &logResponseWriter{
+		ResponseWriter: w,
+		status:         http.StatusOK,
+	}
+
+	start := time.Now()
+	l.handler.ServeHTTP(writerWrapper, r)
+	duration := time.Now().Sub(start)
```
Comment (bboreham) [inline common/http.go]: what is the motivation to include the duration?

File: prog/weaver/main.go [hunk 2]
```diff
@@ -508,7 +512,7 @@ func parseIPAMSeed(ipamSeed string) []mesh.PeerName {
 	return peerNames
 }
 
-func listenAndServeHTTP(httpAddr string, muxRouter *mux.Router) {
+func listenAndServeHTTP(httpAddr string) {
```
Comment (bboreham) [inline prog/weaver/main.go]: this looks like an unrelated change; please put it in a separate commit

**Comment by bboreham** [comment]
@awh as requestor in #2034 do you agree this addresses it?

File: common/http.go (hunk 1 above)
Comment (brb) [inline common/http.go]: In general, I find it useful when debugging failures happening in long chains of calls due to timeouts / timing issues. Also, it can give us a better idea on how long some requests are taking.

File: prog/weaver/main.go (hunk 2 above)
Comment (brb) [inline prog/weaver/main.go]: Done.

**Comment by brb** [comment]
> Please take the unnecessary change out of prog/weaver/http.go

Shall I push such cosmetic changes (trailing whitespace, unnecessary empty lines, OCD, etc.) directly to master or leave them as they are?

File: common/http.go (hunk 1 above)
Comment (bboreham) [inline common/http.go]: Ok, but this points up a new concern: you are logging things after they happen.
In the event of a failure we would prefer to have seen in the logs what the program was about to do.

**Comment by bboreham** [comment]
Minor cosmetic changes can go straight on master.

File: common/http.go (hunk 1 above)
Comment (brb) [inline common/http.go]: Good point, but I assume that in such case (when `l.handler.ServeHTTP` panics) you would be able to reconstruct what happened from a stacktrace.

Anyway, I could add two log statements - before and after. Opinions?

Comment (bboreham) [inline common/http.go]: My preference would be one line before.  I think we can add the duration later if we discover we care.

Comment (brb) [inline common/http.go]: @awh are you OK with that?

Comment (rade) [inline common/http.go]: I don't think we should log the duration *or status code* - failures will generally get logged already anyway.

Which means we don't need `logResponseWriter` and this whole PR will shrink down to a handful of lines. Which is nice.

File: prog/weaver/main.go (hunk 2 above)
Comment (rade) [inline prog/weaver/main.go]: I've actually just fixed this on master - 03afb54c - so this commit should be removed.

File: common/http.go (hunk 1 above)
Comment (brb) [inline common/http.go]: Done.

**Comment by brb** [comment]
PTAL.

File: common/http.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// $METHOD $REQUEST_URI
+func LogHandler(h http.Handler) http.Handler {
+	return &logHandler{handler: h}
+}
+
+// ServeHTTP implements the http.Handler interface.
+func (l *logHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	Log.Debugf("%s %s", r.Method, r.URL.RequestURI())
+	l.handler.ServeHTTP(w, r)
+}
```
Comment (rade) [inline common/http.go]: I'd call this thing `LoggingHTTPHandler` and dispense with *all* the comments above.

File: prog/weaver/main.go [hunk 4]
```diff
@@ -293,7 +293,11 @@ func main() {
 		}
 		router.HandleHTTP(muxRouter)
 		HandleHTTP(muxRouter, version, router, allocator, defaultSubnet, ns, dnsserver)
-		http.Handle("/", muxRouter)
+		var handler http.Handler = muxRouter
+		if logLevel == "debug" {
+			handler = common.LogHandler(muxRouter)
+		}
+		http.Handle("/", handler)
```
Comment (rade) [inline prog/weaver/main.go]: Is there any harm in *always* going via the logging handler? We really don't need to be concerned about efficiency here.

That way all of the above collapses to

    http.Handle("/", common.LoggingHTTPHandler(muxRouter))

Comment (brb) [inline prog/weaver/main.go]: Apart from efficiency, none.

File: common/http.go (hunk 3 above)
Comment (brb) [inline common/http.go]: Done.

**Comment by rade** [comment]
I wonder whether the log lines should have a prefix, e.g. "[http]" or "[api]" or "[http api]", ... to aid in grep'ing.

@bboreham @awh @brb thoughts?

**Comment by bboreham** [comment]
The `LoggingHTTPHandler` constructor could take a parameter which is the prefix, thus removing the obligation on the common part to say something generic.

**Comment by rade** [comment]
@bboreham good idea. Any opinions on what the prefix should be?

**Comment by bboreham** [comment]
Since we're using the same entry-point for router, IPAM and DNS control, it still has to be quite generic.  I would be ok with `"[http]"`.

**Comment by brb** [comment]
Yep, prefixing each log line with "[http]" is a good idea for the sake of grepping.

However, I'm not sure whether it makes sense to extend `LoggingHTTPHandler` by adding the prefix param, because the same instance of the handler is used by all endpoints and it should stay generic.

**Comment by rade** [comment]
>  the same instance of the handler is used by all endpoints and it should stay generic.

Fair point. We can always make the prefix configurable later if we feel the need.
